app/repositories/s3_enrolment_repo.py

```diff
@@ -0,0 +1,107 @@
+import json
+from typing import Any, List
+
+import boto3
+
+from app.config import settings
+from app.domain.entities.enrolment_request import (
+    VALID_STATE_CHANGES,
+    EnrolmentFilters,
+    EnrolmentRequest,
+)
+from app.repositories.enrolment_repo import EnrolmentRepo
+from app.utils.error_handling import handle_s3_errors
+
+
+class S3EnrolmentRepo(EnrolmentRepo):
+    s3: Any
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.s3 = boto3.client("s3", **settings.s3_configuration)
+
+    def get_enrolments(
+        self, enrolment_filters: EnrolmentFilters
+    ) -> List[EnrolmentRequest]:
+        enrolment_filters = enrolment_filters.dict()
+        with handle_s3_errors():
+            enrolment_objects_list = self.s3.list_objects(
+                Bucket=settings.ENROLMENT_BUCKET
+            )
+        enrolments_list = []
+        for enrolments_object in enrolment_objects_list.get("Contents", []):
+            with handle_s3_errors():
+                obj = self.s3.get_object(
+                    Key=f"enrolments/{enrolments_object['Key']}.json",
+                    Bucket=settings.ENROLMENT_BUCKET,
+                )
+            enrolment = json.loads(obj["Body"].read().decode())
+            enrolment = EnrolmentRequest(**enrolment).dict()
+            if filters_match(enrolment, enrolment_filters.copy()):
+                enrolments_list.append(enrolment)
+
+        return enrolments_list
+
+    def get_enrolment_by_id(self, enrolment_request_id: str) -> EnrolmentRequest:
+        with handle_s3_errors():
+            obj = self.s3.get_object(
+                Key=f"enrolments/{enrolment_request_id}.json",
+                Bucket=settings.ENROLMENT_BUCKET,
+            )
+        enrolment = EnrolmentRequest(**json.loads(obj["Body"].read().decode()))
+        return enrolment
+
+    def update_enrolment_state(
+        self, enrolment: EnrolmentRequest, new_state: str
+    ) -> EnrolmentRequest:
+        new_enrolment = enrolment.dict()
+        new_enrolment.update({"state": new_state})
+        new_enrolment = EnrolmentRequest(**new_enrolment)
+        with handle_s3_errors():
+            self.s3.put_object(
+                Body=bytes(new_enrolment.json(), "utf-8"),
+                Key=f"enrolments/{new_enrolment.enrolment_id}.json",
+                Bucket=settings.ENROLMENT_BUCKET,
+            )
+        return new_enrolment
+
+
+def filters_match(enrolment: dict, enrolment_filters: dict):
```

Comment:
glad it helped :') I was just thinking if admin wants to filter enrl requests by competency or standards too?

Reply:
Yes, thanks :) I think you're confusing enrolment authorisation with courses. We already might get courses by these filters. And then, we can use course IDs to filter enrolment authorisations.